6 спринт

FanEm opened the pull request:
Добавлена работа с сетью, список фильмов загружается из API IMDb. 
Обработаны состояния приложения при загрузке фильмов и получения ошибок.

FanEm committed:
6 спринт
Добавлена работа с сетью, список фильмов загружается из API IMDb.
Обработаны состояния приложения при загрузке фильмов и получения ошибок.

Helvetii951 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -65,7 +71,31 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+    }
+
     // MARK: - Private functions
+    private func showLoadingIndicator() {
+        activityIndicator.isHidden = false
+        activityIndicator.startAnimating()
+    }
+
+    private func hideLoadingIndicator() {
+        activityIndicator.isHidden = true
+        activityIndicator.stopAnimating()
+    }
```
Рекомендую посмотреть на свойство [hidesWhenStopped](https://developer.apple.com/documentation/uikit/uiactivityindicatorview/1622837-hideswhenstopped) 
у `activityIndicator`, и использовать его вместо ручного управления свойством `isHidden`. 

Когда ты будешь вызывать `startAnimating` – индикатор будет появляться автоматически, соответственно на `stopAnimating` – скрываться, а свойство `isHidden` больше не понадобится, индикатор все сделает за тебя. 

**Важно:** это свойство настраивают всего раз раз: кодом во `viewDidLoad()` или аналогичной галочкой у элемента на сториборде.
А строки `activityIndicator.isHidden = false(true)` нужно будет убрать

FanEm replied:
Спасибо, исправил

Helvetii951 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -52,14 +60,60 @@ final class QuestionFactory: QuestionFactoryProtocol {
+            guard let movie = self.movies[safe: index] else { return }
+
+            var imageData = Data()
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("Failed to load image")
+            }
```
Обращу внимание, что игра состоит из вопроса о рейтинге, на который отвечает пользователь, и контекста о фильме в виде постера(картинка). 
Без постера, пользователю будет затруднительно ответить на вопрос о рейтинге фильма верно.

В этом месте, функция не завершает свою работу и идет формировать вопрос без постера. 
В таком случае, бизнес-логика приложения будет нарушена, потому что пользователь не сможет ответить корректно, ведь изображения нет.
Это состояние стоит обработать, например, с помощью алерта. 
Не забудь, что алерт будет обновлять UI и это тоже нужно учесть (вернуть на main), ровно также, как и прерывание функции (сделать return)

FanEm replied:
Исправлено

Helvetii951 commented on MovieQuiz/Services/NetworkClient.swift:
```diff
@@ -0,0 +1,34 @@
+            }
+
+            guard let data else { return }
+            handler(.success(data))
```
`guard let data else { return }`. Именно в этом контексте `data` скорее всего будет всегда. Но. в guard не вызывается handler перед `return`. Т.е. он никогда не будет вызван. Очень рекомендую добавить вызов `handler`a с новой ошибкой NetworkError.emptyData

FanEm replied:
Исправлено

FanEm committed:
Добавлен лоадинг в момент загрузки постера
Добавлена обработка ошибки загрузки постера
Добавлена обработка ошибки получения данных
Написан метод для создания разнообразных вопросов
